### ISSUE TITLE
Tab: focused element and active tab id can go out of sync

### ISSUE THREAD

**Issue by ciampo** [ariakit/ariakit#4213]
### Current behavior

Keyboard focus and active tab id can go out of sync if `Tab` receives a controlled update while the keyboard focus is on a tab within the tablist.

https://github.com/user-attachments/assets/61768a61-22ff-40a7-91d4-847e1c320a6f

Note: this behavior can be reproduced regardless of the `selectOnMove` preference.

### Steps to reproduce the bug

1. Open https://stackblitz.com/edit/xpiaoy?file=tab%2Findex.tsx&theme=dark
2. When the example loads, tab into the tablist and press the right arrow key to select the "Vegetables" tab
3. Wait for the timeout to run, applying a controlled update and selecting the "Meat" tab
4. At this point, the Vegetables tab has keyboard focus, but the Meat tab is selected
5. Press the right arrow key.
6. The "Fruit" tab gets selected

### Expected behavior

I agree that we shouldn't change the focused item (and I think that the behavior described at point 4 is correct), but as I user I would be confused by what happens at point 6: I'd expect the "Meat" tab to receive keyboard focus, and not the "Fruit" tab.

In other words, I'd expect the active tab to stay the same as the focused tab, instead of following the selected tab.

### Workaround

Manually re-set the active ID if it goes out of sync.

```jsx
  const { items, activeId } = Ariakit.useStoreState(tabStore);
  const { setActiveId } = tabStore;

  useEffect(() => {
    requestAnimationFrame(() => {
      const focusedElement = items?.[0]?.element?.ownerDocument.activeElement;

      if (
        !focusedElement ||
        !items.some((item) => focusedElement === item.element)
      ) {
        // Return early if no tabs are focused.
        return;
      }

      // If, after ariakit re-computes the active tab, that tab doesn't match
      // the currently focused tab, then we force an update to ariakit to avoid
      // any mismatches, especially when navigating to previous/next tab with
      // arrow keys.
      if (activeId !== focusedElement.id) {
        setActiveId(focusedElement.id);
      }
    });
  }, [activeId, items, setActiveId]);
```

**Comment by benrodrs** [comment]
We update the `activeId` state when `selectedId` changes here:

https://github.com/ariakit/ariakit/blob/b693d4aa204702c1ca85c10babfd63885f7a4a1b/packages/ariakit-core/src/tab/tab-store.ts#L109-L125

I think this makes sense when no tab has DOM focus, so tabbing into the tab list will focus on the newly selected tab. However, we should probably avoid this if another item is currently focused.

**Comment by benrodrs** [comment]
> receives a controlled update

Also, I can confirm this issue occurs with uncontrolled `tabStore.setSelectedId('meat')` too. The only scenario where it consistently works is with uncontrolled `tabStore.select('meat')`, although the DOM focus moves to the selected tab.

**Comment by diegohaz** [comment]
We should decide if the DOM focus should move to the selected tab. For consistency, I'd always move the focus. Otherwise, the UX becomes awkward when `selectOnMove` is `true` (default).

@ciampo Do you have a specific example where this happens? I mean, a real scenario where the selected tab changes without user action while they're focusing on another tab.

**Comment by ciampo** [comment]
The workaround described above was:
- introduced in Gutenberg in https://github.com/WordPress/gutenberg/pull/56658
- tweaked in https://github.com/WordPress/gutenberg/pull/57696

**Comment by diegohaz** [comment]
Thanks! If I understand correctly, the selected tab changes shortly after the user tabs from the block to the tab list. In this situation, I'd expect the focus to move to the selected tab, as it would if the tab were selected synchronously.

**Comment by ciampo** [comment]
> In this situation, I'd expect the focus to move to the selected tab, as it would if the tab were selected synchronously.

At the time, we agreed that moving focus to a different tab would be disruptive ([example](https://github.com/WordPress/gutenberg/pull/56658#issuecomment-1836252047)).

Just to understand, what are the reasons why you would push back against the suggested behavior (Ie. active id doesn't sync with new selected id if keyboard focus is in the composite widget) ?

**Comment by diegohaz** [comment]
> Just to understand, what are the reasons why you would push back against the suggested behavior (Ie. active id doesn't sync with new selected id if keyboard focus is in the composite widget) ?

I understand that in the scenario you mentioned, the focus would change so quickly that it would appear as if the selected tab was the initially focused element, which is what you would expect if the tab were selected synchronously. I need to test this, but I believe that in this situation, screen readers might not announce the element focused for a fraction of a second.

This is theoretical, and there might be other use cases I'm not aware of.

**Comment by ciampo** [comment]
Also, another question:

> Thanks! If I understand correctly, the selected tab changes shortly after the user moves focus from the block to the tab list. In this situation, I'd expect the focus to move to the selected tab, as it would if the tab were selected synchronously.

If you expect focus to move to the selected tab, wouldn't that be something that Ariakit should implement?

What I'm trying to say is that, currently, if `selectedId` changes while a tab is focused, keyboard focus and `activeId` go out of sync. I would assume that it's in the interest of ariakit to prevent that from happening, meaning that either the `activeId` is updated to match keyboard focus, or the keyboard focus is moved to element associated with the active id

**Comment by diegohaz** [comment]
> If you expect focus to move to the selected tab, wouldn't that be something that Ariakit should implement?

Yes, but since we need to make a design decision, I wanted to understand the use cases before implementing anything. This also helps us create real-world examples with automated tests.

---

I guess this is what you meant, right?

https://github.com/user-attachments/assets/bf6cf0ce-52b8-4952-bc2b-d9f76abec2b7

In view mode, pressing <kbd>Tab</kbd> deselects the block and updates the selected tab. It seems that focus follows the tab selection in this case.

However, this behavior seems problematic even if we implement a solution in Ariakit, which would mimic the current functionality in Gutenberg if we decide to focus on the selected tab. In this scenario, a keyboard user cannot access the block tab panel while in view mode.

**Comment by ciampo** [comment]
I think that we have to differentiate between:

- the [general problem](https://github.com/ariakit/ariakit/issues/4213#issuecomment-2422571158), which IMO needs addressing (related unit test that we currently have in `@wordpress/components`)
- the [specific Gutenberg issue](https://github.com/ariakit/ariakit/issues/4213#issuecomment-2416150887), where I agree with you that we could handle the situation better in "consumer-land", regardless the behaviour of `Tabs`

**Comment by diegohaz** [comment]
Just to clarify the steps here:

-  [x] Acknowledge this is a bug that needs to be fixed in the library.
-  [x] Decide whether to move focus or prevent `activeId` from updating (seems almost settled on the former?).
-  [x] Identify valid use cases to help with the decision above and serve as examples with automated tests. This also helps document directly in the codebase _why_ something works in a specific way.

**Comment by ciampo** [comment]
Thank you for clarifying (and feel free to hoist the checklist to the issue's description if you'd like).

> Decide whether to move focus or prevent activeId from updating (seems almost settled on the former?).

I'm not fully convinced about which option is the best (and therefore I don't have a strong opinion), but any option would be better than current behavior.

>  Identify valid use cases to help with the decision above and serve as examples with automated tests. This also helps document directly in the codebase why something works in a specific way.

[This use case](https://github.com/ariakit/ariakit/issues/4213#issuecomment-2422690123) seems legit. I'll see if I can think of more

**Comment by ciampo** [comment]
@diegohaz Thank you for fixing this!
